hooks/process_items.py

```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sgtk
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+class ProcessItems(HookBaseClass):
+    """
+    This hook defines methods that operates on top-level publish items during
+    the publishing process.
+    """
+
+    def prepare(self, items):
```

Comment:
any reason we pass a list? would it be better to have a yield-style pattern we we effectively call this method several times, for each item? The current implementation prevents us from doing certain things in the future, say if we wanted to compute stuff remotely or something, or run in parallel. 

It kind of breaks the pattern we have with plugins where we run each piece of compute in a very decoupled, decentralized fashion.

Can you foresee workflows where a list of all the items would be needed as input to a process?

Reply:
This was done on a client request. 

>have a new hook which gets run after all the items get published, the main reason for this is so we can have the items gather the scene publish data and have a global hook batch submit it all to the farm/SG Database so we don't do all the exporting and publishing locally

I can reach out to the client to see if they envisage any problems with running it in a decoupled way/per plugin?

Also, what would the approach be if for example you created a bunch of versions from each item, and then you wanted to add them to the same playlist. I could see a final global hook being useful for this purpose, but in a decoupled approach would it be that the first items would run and create the playlist and pass it on to the others so they could attach? Or a playlist name would be decided at the beginning of the collector, and each would simply find and attach its version to it? (maybe this is going off track, but I was just trying to think of times when you would require a global approach.)

Reply:
hey phil, i talked to josh about it yesterday evening and i think he is tweaking the code right now. The conclusion is that we keep the proposed implementation in this PR, and add some more hooks! So i don't think we need to reach out to the client!